### PR TITLE
🌱Add condition indicating deletion timestamp is set

### LINKED ISSUE #1229: Condition indicating that HetznerBareMetalHost is deleting
/kind feature

**Describe the solution you'd like**
The deletion process of a HetznerBareMetalHost needs some time. We can introduce a condition / a reason that indicates this and as such is printed in the printcolumns of the object. This makes it easier for users because they have an immediate feedback without having to check the deletion timestamp of the object.

### PR DESCRIPTION
**Pull request by yrs147**
**What this PR does / why we need it**:
Add condition indicating deletion timestamp is set
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1229
**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

### CODE REVIEW

File: api/v1beta1/conditions_const.go [hunk 1]
```diff
@@ -200,6 +200,13 @@ const (
 	HostAssociateFailedReason = "HostAssociateFailed"
 )
 
+const (
```
[the remark this answers is not in the document]
Comment (janiskemper) [reply api/v1beta1/conditions_const.go:203]: what about "HostReadyCondition"? https://github.com/syself/cluster-api-provider-hetzner/blob/main/api/v1beta1/conditions_const.go#L103

[the remark this answers is not in the document]
Comment (yrs147) [reply api/v1beta1/conditions_const.go:203]: A host can be  `Available` but not `Ready` , in which case marking `HostReadyCondition` to false won't be correct imo , thats why I suggested `HostAvailiableCondition`

[the remark this answers is not in the document]
Comment (janiskemper) [reply api/v1beta1/conditions_const.go:203]: sorry I don't get what you mean tbh :D If you don't want to elaborate via text we can just talk in the next days

File: controllers/hetznerbaremetalhost_controller.go [hunk 2]
```diff
@@ -170,6 +170,18 @@ func (r *HetznerBareMetalHostReconciler) reconcileSelectedStates(ctx context.Con
...
 		var needsUpdate bool
 		if !bmHost.DeletionTimestamp.IsZero() && bmHost.Spec.ConsumerRef == nil {
+			conditions.MarkTrue(
+				bmHost,
+				infrav1.DeletionInProgressCondition,
+			)
+			conditions.MarkFalse(
+				bmHost,
+				infrav1.CredentialsAvailableCondition,
+				infrav1.HetznerSecretUnreachableReason,
```
Comment (guettli) [inline controllers/hetznerbaremetalhost_controller.go:180]: is `HetznerSecretUnreachableReason` the correct reason?

Comment (yrs147) [reply controllers/hetznerbaremetalhost_controller.go:180]: not sure , what do would you suggest it to be ?

Comment (guettli) [reply controllers/hetznerbaremetalhost_controller.go:180]: I would set it like below:

```
			conditions.MarkFalse(
				bmHost,
				infrav1.DeletionInProgressCondition,
				infrav1.DeletionInProgressReason,
				clusterv1.ConditionSeverityInfo,
				"Deletion in progress",
			)
```

Comment (yrs147) [reply controllers/hetznerbaremetalhost_controller.go:180]: should I create a `CredentialsAvailableReason` for this then ?
cc @janiskemper

File: controllers/hetznerbaremetalhost_controller.go [hunk 3]
```diff
@@ -188,6 +196,15 @@ func (r *HetznerBareMetalHostReconciler) reconcileSelectedStates(ctx context.Con
 		// Handle StateDeleting
 	case infrav1.StateDeleting:
 		if !utils.StringInList(bmHost.Finalizers, infrav1.BareMetalHostFinalizer) {
+			conditions.MarkFalse(
```
Comment (janiskemper) [inline controllers/hetznerbaremetalhost_controller.go:199]: I think you can remove this. Right now it doesn't have any effect (see that you don't update the host after this). You could, if anything, add it to the code below, but also there you remove the finalizer already. After the finalizer is removed, the object will be deleted very quickly. Therefore, we probably don't need a condition here at all.

File: controllers/hetznerbaremetalhost_controller.go [hunk 4]
```diff
@@ -170,6 +170,14 @@ func (r *HetznerBareMetalHostReconciler) reconcileSelectedStates(ctx context.Con
 	case infrav1.StateNone:
 		var needsUpdate bool
 		if !bmHost.DeletionTimestamp.IsZero() && bmHost.Spec.ConsumerRef == nil {
+			conditions.MarkFalse(
```
Comment (janiskemper) [inline controllers/hetznerbaremetalhost_controller.go:173]: you should set this condition independent of Spec.ConsumerRef. So just split up the if statements to set the condition always when the deletion timestamp is not nil

**Review by janiskemper — CHANGES_REQUESTED** [review]

File: controllers/hetznerbaremetalhost_controller.go [hunk 5]
```diff
@@ -180,19 +180,26 @@ func (r *HetznerBareMetalHostReconciler) reconcileSelectedStates(ctx context.Con
 	switch bmHost.Spec.Status.ProvisioningState {
 	// Handle StateNone: check whether needs to be provisioned or deleted.
 	case infrav1.StateNone:
-		var needsUpdate bool
-		if !bmHost.DeletionTimestamp.IsZero() && bmHost.Spec.ConsumerRef == nil {
-			bmHost.Spec.Status.ProvisioningState = infrav1.StateDeleting
-			needsUpdate = true
+		if !bmHost.DeletionTimestamp.IsZero() {
+			if !conditions.Has(bmHost, infrav1.HostReadyCondition) {
```
Comment (janiskemper) [inline controllers/hetznerbaremetalhost_controller.go:184]: you shouldn't do this check here. You should set it independently of whether it exists or not. You even have to overwrite it. 

What I meant earlier was that you might still keep needsUpdate and then need to check whether the condition you set actually changed something in the object or not. But for that you need to check the reason as well, not only that the condition exists. 

I would just remove this if statement and always set the condition on false and you are good to go!

**Review by janiskemper — CHANGES_REQUESTED** [review]

**Review by janiskemper — COMMENTED** [review]
LGTM. Thank you! Please squash your commits now

File: pkg/services/baremetal/host/host.go [hunk 6]
```diff
@@ -91,6 +91,29 @@ func NewService(s *scope.BareMetalHostScope) *Service {
 
 // Reconcile implements reconcilement of HetznerBareMetalHosts.
 func (s *Service) Reconcile(ctx context.Context) (result reconcile.Result, err error) {
+	// Fetch the Hetzner bare metal host instance.
```
Comment (janiskemper) [inline pkg/services/baremetal/host/host.go:94]: the host is in the scope (s.scope)

File: pkg/services/baremetal/host/host.go [hunk 7]
```diff
@@ -91,6 +91,29 @@ func NewService(s *scope.BareMetalHostScope) *Service {
...
+	}
+	if !bmHost.DeletionTimestamp.IsZero() {
+		conditions.MarkFalse(
+			bmHost,
+			infrav1.HostReadyCondition,
+			infrav1.DeletionInProgressReason,
+			clusterv1.ConditionSeverityInfo,
+			"Host is not ready because it is being deleted",
+		)
+		conditions.SetSummary(bmHost)
```
Comment (janiskemper) [inline pkg/services/baremetal/host/host.go:114]: "setsummary" should be done anyway in the defer below

File: pkg/services/baremetal/host/host.go [hunk 8]
```diff
@@ -91,6 +91,29 @@ func NewService(s *scope.BareMetalHostScope) *Service {
...
+		Namespace: s.scope.HetznerBareMetalHost.Namespace,
+		Name:      s.scope.HetznerBareMetalHost.Name,
+	}, bmHost)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get HetznerBareMetalHost : %w", err)
+	}
+	if !bmHost.DeletionTimestamp.IsZero() {
```
Comment (janiskemper) [inline pkg/services/baremetal/host/host.go:106]: you need to put this code after the "oldHost" is defined. The oldHost is the "original" state of the host and all changes thereafter are updated via the "Update" API call below

**Review by janiskemper — CHANGES_REQUESTED** [review]

**Review by janiskemper — COMMENTED** [review]
can you revert all changes to hetznerbaremetalhost_controller? I think those changes are not relevant anymore

**Review by janiskemper — COMMENTED** [review]
thanks! This looks good now. You can squash all commits and do one small change:
use the severity "warning" instead of "info". I'm a bit afraid that it wouldn't be shown to users otherwise

**Review by janiskemper — APPROVED** [review]

**Pull request merged**